File: src/tools/pmd-cpd/eval-repos/synthetic/typescript/no_duplication.ts
/**
 * TypeScript file with no code duplication - all functions are unique.
 */

interface User {
    id: number;
    name: string;
    email: string;
    active: boolean;
}

function calculateFibonacci(n: number): number {
    if (n <= 1) return n;
    let a = 0, b = 1;
    for (let i = 2; i <= n; i++) {
        [a, b] = [b, a + b];
    }
    return b;
}

function isPrime(num: number): boolean {
    if (num < 2) return false;
    if (num === 2) return true;
    if (num % 2 === 0) return false;
    for (let i = 3; i <= Math.sqrt(num); i += 2) {
        if (num % i === 0) return false;
    }
    return true;
}

function binarySearch<T>(arr: T[], target: T): number {
    let left = 0, right = arr.length - 1;
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        if (arr[mid] === target) return mid;
        if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

class DataProcessor<T extends number> {
    private data: T[];
    private processed: boolean = false;

    constructor(data: T[]) {
        this.data = data;
    }

    normalize(): number[] {
        if (!this.data.length) return [];
        const min = Math.min(...this.data);
        const max = Math.max(...this.data);
        if (max === min) return this.data.map(() => 0.5);
        return this.data.map(x => (x - min) / (max - min));
    }

    computeStatistics(): { mean: number; median: number; std: number } {
        if (!this.data.length) return { mean: 0, median: 0, std: 0 };
        const n = this.data.length;
        const mean = this.data.reduce((a, b) => a + b, 0) / n;
        const sorted = [...this.data].sort((a, b) => a - b);
        const median = n % 2 ? sorted[Math.floor(n / 2)] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2;
        const variance = this.data.reduce((sum, x) => sum + (x - mean) ** 2, 0) / n;
        return { mean, median, std: Math.sqrt(variance) };
    }
}

export { User, calculateFibonacci, isPrime, binarySearch, DataProcessor };

File: src/tools/pmd-cpd/eval-repos/synthetic/rust/no_duplication.rs
//! Rust file with no code duplication - all functions are unique.

/// Calculate the nth Fibonacci number.
pub fn calculate_fibonacci(n: u64) -> u64 {
    if n <= 1 {
        return n;
    }
    let (mut a, mut b) = (0u64, 1u64);
    for _ in 2..=n {
        let temp = a + b;
        a = b;
        b = temp;
    }
    b
}

/// Check if a number is prime.
pub fn is_prime(num: u64) -> bool {
    if num < 2 {
        return false;
    }
    if num == 2 {
        return true;
    }
    if num % 2 == 0 {
        return false;
    }
    let sqrt = (num as f64).sqrt() as u64;
    for i in (3..=sqrt).step_by(2) {
        if num % i == 0 {
            return false;
        }
    }
    true
}

/// Perform binary search on a sorted array.
pub fn binary_search(arr: &[i32], target: i32) -> Option<usize> {
    let (mut left, mut right) = (0, arr.len());
    while left < right {
        let mid = (left + right) / 2;
        if arr[mid] == target {
            return Some(mid);
        }
        if arr[mid] < target {
            left = mid + 1;
        } else {
            right = mid;
        }
    }
    None
}

/// Data processor for numerical data.
pub struct DataProcessor {
    data: Vec<f64>,
    processed: bool,
}

impl DataProcessor {
    pub fn new(data: Vec<f64>) -> Self {
        Self { data, processed: false }
    }

    /// Normalize the data to 0-1 range.
    pub fn normalize(&self) -> Vec<f64> {
        if self.data.is_empty() {
            return vec![];
        }
        let min = self.data.iter().cloned().fold(f64::INFINITY, f64::min);
        let max = self.data.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
        if (max - min).abs() < f64::EPSILON {
            return vec![0.5; self.data.len()];
        }
        self.data.iter().map(|x| (x - min) / (max - min)).collect()
    }

    /// Compute basic statistics on the data.
    pub fn compute_statistics(&self) -> Statistics {
        if self.data.is_empty() {
            return Statistics::default();
        }
        let n = self.data.len() as f64;
        let mean = self.data.iter().sum::<f64>() / n;

        let mut sorted = self.data.clone();
        sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
        let median = if sorted.len() % 2 != 0 {
            sorted[sorted.len() / 2]
        } else {
            (sorted[sorted.len() / 2 - 1] + sorted[sorted.len() / 2]) / 2.0
        };

        let variance = self.data.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / n;
        Statistics { mean, median, std: variance.sqrt() }
    }
}

#[derive(Default, Debug)]
pub struct Statistics {
    pub mean: f64,
    pub median: f64,
    pub std: f64,
}
